app/assets/javascripts/pages/referenceWidget.js

```diff
@@ -0,0 +1,229 @@
+      });
+    
+    this.options.afterRender();
+  };
+
+  ///////////////////////////
+  // TYPEAHEAD COMPONENTS ///
+  //////////////////////////
+  
+
+  /**
+   * Prepares Bloodhound search with provided references data
+   * @param {Array[Referencs]} data
+   * @returns {Bloodhound} 
+   */
+  function documentBloodhound(data) {
+    return new Bloodhound({
+      local: data,
+      identify: function(doc) { return doc.id; },
+      queryTokenizer: function(q) {
+	return Bloodhound.tokenizers.whitespace(q.toLowerCase());
+      },
+      datumTokenizer: function(d) {
+	return flatten(
+	  [d.name, d.url]
+	    .map(function(x) {
+	      return Bloodhound.tokenizers.whitespace(x.toLowerCase());
+	    })
+	);
+      }
+    });
+  }
+  
+  /**
+   * Provides a configuartion object for typeahead
+   * Requires `this.documents` to be set
+   * @param {Array} references
+   * @returns {Object} Configuration for typeahead 
+   */
+  ExistingReferenceWidget.prototype._typeaheadConfig = function() {
+    if (!Array.isArray(this.documents)) {
+      throw 'Documents are missing or invalid :(';
+    }
+    
+    return {
+      name: 'references',
+      source: documentBloodhound(this.documents),
+      displayKey: 'name',
+      templates: {
+	empty: '<span class="reference-suggestion"><span class="reference-empty">No matches found</span></span>',
+	suggestion: suggestion
+      }
+    };
+  };
+  
+
+  /**
+   * Retrives recent documents via an ajax call to /references/recent
+   * Sets this.documents if successful.
+   * @returns {Promise}
+   */
+  ExistingReferenceWidget.prototype.getDocs = function() {
+    var self = this;
+    var url = this._recentReferencesUrl();
+    
+    return new Promise(function(resolve, reject){
+      $.getJSON(url)
+	.done(function(data) {
+	  self.documents = data;
+	  resolve(data);
+	})
+	.fail(function(){
+	  reject('failed to get references from /references/recent');
+	});
+    });
+  };
+
+
+  /**
+   * Url for request
+   * @returns {String} 
+   */
+  ExistingReferenceWidget.prototype._recentReferencesUrl = function() {
+    var params =  $.param({
+      "entity_ids":  this.entityIds,
+      "per_page": REFERENCES_PER_PAGE,
+      "exclude_type": 'fec'
+    });
+
+    return REQUEST_URL + '?' + params;
+  };
+
+
+  ///////////////////
+  // DOM ELEMENTS ///
+  //////////////////
+    
+  /**
+   * Input with id ref-widget-typeahead
+   * @returns {<input>} 
+   */
+  ExistingReferenceWidget.prototype._typeaheadInput = function() {
+    return $('<input>', {
+      "type": 'text',
+      "placeholder": 'Select an existing reference',
+      "id": ExistingReferenceWidget.TYPEAHEAD_INPUT_ID,
+      "class": 'reference-typeahead'
+    });
+  };
+  
+
+  var referenceVisitLink = [
+    '<div class="reference-visit-link">',
+    '<a href="{{url}}" target="_blank">',
+    '<span class="glyphicon glyphicon-new-window"></span>',
+    '</a></div>'
+  ].join('');
+
+  
+  var suggestionRender = Hogan.compile(
```

Comment:
curious what motivated the move to use hogan/handlebar templates instead of pseudo-components (ie: just functions with arguments) as in other jquery.

just trying something different for a change? do you want to move to using hogan more throughout the codebase?

Reply:
we use hogan//handlebar in other places in our javascript as well. sometimes i find it easier than using jQuery(), but  i don't really have a preference...at some point i want to replace hogan with mustache.js because hogan isn't being maintained...